### ISSUE TITLE
Unable to set components version using craftctl set version

### ISSUE THREAD

**Issue by stephensp** [canonical/snapcraft#5102]
### What needs to get done

component versions need to be settable during override-build , just like a snap.  Currently it fails with the following error

RuntimeError: 'override-build' in part 'nvidia-550-ko-comp' executed an invalid control API call: variable 'version' can be set only once.

### Why it needs to get done

components need to be able to dynamically set their version based on revision numbers and other system settings. This is not possible if the version has to be hardcoded into the version field.

**Comment by sergiusens** [comment]
This is because the part is decoupled from the component, this would need to be a new craftctl command specific to the component that takes a component name and the desired version, or overload the current set version command to use a key=value (component=version) like syntax.

**Comment by sergiusens** [comment]
Of course, then adopt-info would need to be considered for each component

**Comment by cmatsuoka** [comment]
Maybe `craftctl set version(<component_name>)=<value>`?

**Comment by mr-cal** [comment]
Or @lengau's suggestion `craftctl set comp1.version=<value>`.

After discussing with @sergiusens and @cmatsuoka, this is unlikely to be completed this cycle.  Claudio is going to discuss this proposal with Gustavo tomorrow.

**Comment by mr-cal** [comment]
We've [spec'd](https://docs.google.com/document/d/1s58743BPMb_TECU_epvJnbsxIomglcbJ3YvqKpZ4pzw/edit?tab=t.0) and prototyped this work ([snapcraft](https://github.com/canonical/snapcraft/compare/main...work/CRAFT-3605-craftctl-partition-prototype), [craft-application](https://github.com/canonical/craft-application/compare/main...work/CRAFT-3605-craftctl-partition-prototype), [craft-parts](https://github.com/canonical/craft-parts/compare/main...work/CRAFT-3605-craftctl-partition-prototype)).

The work to implement this hasn't been scheduled. It is a good candidate for next cycle.

Related: #5240
